nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/CustomKerberosLogin.java

```diff
@@ -0,0 +1,245 @@
+        return subject;
+    }
+
+    @Override
+    public String serviceName() {
+        return serviceName;
+    }
+
+    private synchronized void checkTGTAndReLogin() throws LoginException {
+        final KerberosTicket tgt = getTGT();
+        if (tgt == null) {
+            log.info("[Principal={}]: TGT was not found, performing login", principal);
+            reLogin();
+            return;
+        }
+
+        if (System.currentTimeMillis() < getRefreshTime(tgt)) {
+            log.debug("[Principal={}]: TGT was found, but has not reached expiration window", principal);
+            return;
+        }
+
+        try {
+            tgt.refresh();
+            log.info("[Principal={}]: TGT refreshed", principal);
+            getRefreshTime(tgt);
+        } catch (RefreshFailedException rfe) {
+            log.warn("[Principal={}]: TGT refresh failed, will attempt relogin", principal);
+            log.debug("", rfe);
+            reLogin();
+        }
+    }
+
+    private static String getServiceName(Map<String, ?> configs, String contextName, Configuration configuration) {
+        List<AppConfigurationEntry> configEntries = Arrays.asList(configuration.getAppConfigurationEntry(contextName));
+        String jaasServiceName = JaasContext.configEntryOption(configEntries, JaasUtils.SERVICE_NAME, null);
+        String configServiceName = (String) configs.get(SaslConfigs.SASL_KERBEROS_SERVICE_NAME);
+        if (jaasServiceName != null && configServiceName != null && !jaasServiceName.equals(configServiceName)) {
+            String message = String.format("Conflicting serviceName values found in JAAS and Kafka configs " +
+                    "value in JAAS file %s, value in Kafka config %s", jaasServiceName, configServiceName);
+            throw new IllegalArgumentException(message);
+        }
+
+        if (jaasServiceName != null)
+            return jaasServiceName;
+        if (configServiceName != null)
+            return configServiceName;
+
+        throw new IllegalArgumentException("No serviceName defined in either JAAS or Kafka config");
+    }
+
+
+    private long getRefreshTime(final KerberosTicket tgt) {
+        long start = tgt.getStartTime().getTime();
+        long expires = tgt.getEndTime().getTime();
+
+        log.debug("[Principal={}]: TGT valid starting at: {}", principal, tgt.getStartTime());
+        log.debug("[Principal={}]: TGT expires: {}", principal, tgt.getEndTime());
+        log.debug("[Principal={}]: TGT renew until: {}", principal, tgt.getRenewTill());
+
+        return start + (long) ((expires - start) * ticketRenewWindowFactor);
+    }
+
+    private KerberosTicket getTGT() {
+        Set<KerberosTicket> tickets = subject.getPrivateCredentials(KerberosTicket.class);
+        for (KerberosTicket ticket : tickets) {
+            KerberosPrincipal server = ticket.getServer();
+            if (server.getName().equals("krbtgt/" + server.getRealm() + "@" + server.getRealm())) {
```

Comment:
This would be easier to follow if the value for comparison were formed on a separate line:
```suggestion
            final String expectedServerName = String.format("krbtgt/%s@%s", server.getRealm(), server.getRealm());
            if (server.getName().equals(expectedServerName)) {
```